File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_audit.php
<?php
declare(strict_types=1);

// Routes: Play campaign actor audit trail
// ---------------------------------------------------------------------------

function next_audit_sequence(PDO $db, string $campaignId): int {
    $stmt = $db->prepare('SELECT COALESCE(MAX(seq), 0) AS max_seq FROM play_campaign_audit_events WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    return (int)$stmt->fetch(PDO::FETCH_ASSOC)['max_seq'] + 1;
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/audit-events$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isOwner = $actor['username'] === $campaign['owner'];
    if (!$isOwner && !is_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('must be a campaign member to create audit events');
    }

    $body = read_json_body();
    if ($body === null || !isset($body['kind'], $body['correlation_id'])
        || !is_string($body['kind']) || $body['kind'] === ''
        || !is_string($body['correlation_id']) || $body['correlation_id'] === '') {
        bad_request();
    }

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_audit_events WHERE campaign_id = ? AND correlation_id = ?');
    $stmt->execute([$campaignId, $body['correlation_id']]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('correlation_id already used in this campaign');
    }

    $seq = next_audit_sequence($db, $campaignId);
    $entry = [
        'kind' => $body['kind'],
        'actor' => $actor['username'],
        'role' => $isOwner ? 'DM' : 'player',
        'timestamp' => $seq,
        'correlation_id' => $body['correlation_id'],
    ];

    $stmt = $db->prepare('INSERT INTO play_campaign_audit_events (campaign_id, seq, correlation_id, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $seq, $body['correlation_id'], json_encode($entry)]);

    send_json($entry, 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/audit-events$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if ($actor['username'] !== $campaign['owner']) {
        forbidden('only the campaign owner may read the audit trail');
    }

    $stmt = $db->prepare('SELECT data FROM play_campaign_audit_events WHERE campaign_id = ? ORDER BY seq ASC');
    $stmt->execute([$campaignId]);
    $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
    $entries = array_map(function ($r) {
        return json_decode($r['data'], true);
    }, $rows);

    send_json(['entries' => array_values($entries)]);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/play_audit.rb
# frozen_string_literal: true

# Route definitions for campaign-scoped actor audit events.

# --- Create audit event ---

post '/v1/play/campaigns/:id/audit-events' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  actor = authenticate_actor!
  username = actor[:username]
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_access!(campaign, username)

  body = parse_json_body
  validate_audit_event_body!(body)

  kind = body['kind']
  correlation_id = body['correlation_id']

  json_error(409, 'duplicate correlation_id') if Storage.audit_event_exists?(campaign_id, correlation_id)

  role = audit_role_for(campaign, username)
  entry = Storage.create_audit_event(campaign_id, kind, username, role, correlation_id)

  status 201
  JSON.dump(entry)
end

# --- Read audit events ---

get '/v1/play/campaigns/:id/audit-events' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  actor = authenticate_actor!
  username = actor[:username]
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  entries = Storage.load_audit_events(campaign_id)

  JSON.dump(entries: entries)
end
